core/src/autogluon/core/searcher/bayesopt/gpautograd/gp_regression.py

```diff
@@ -83,45 +81,38 @@ def __init__(
     def states(self) -> Optional[List[GaussProcPosteriorState]]:
         return self._states
 
-    @property
-    def ctx(self):
-        return self._ctx
-
     def _create_lbfgs_arguments(self, X, Y):
         """
-        Prepare the objective, parameters and the gradient arguments for
-        L-BFGS-B
+        Creates SciPy optimizer objective and param_dict for criterion
+        function.
 
-        :param X: data matrix X of size (n, d) (type mx.nd)
-        :param Y: vector of targets of size (n, 1) (type mx.nd)
+        :param X: data matrix X of size (n, d)
+        :param Y: vector of targets of size (n, 1)
+        :return: scipy_objective, param_dict
         """
-
-        def executor():
+        param_dict = {
+            param.name: param
+            for param in self.likelihood.collect_params().values()}
+        # Wraps param_dict, conversion to/from flat vector:
+        param_converter = ParamVecDictConverter(param_dict)
+
+        def executor(param_vec):
+            param_converter.from_vec(param_vec)  # Assign param_dict
             if self._debug_writer is not None:
                 self._debug_writer.store_args(
                     self.likelihood.collect_params().values(), X, Y,
                     self.likelihood.param_encoding_pairs())
-            with autograd.record():
-                objective_nd = negative_log_posterior(self.likelihood, X, Y)
-            objective_np = objective_nd.asscalar()
+            objective = negative_log_posterior(self.likelihood, X, Y)
             if self._debug_writer is not None:
-                self._debug_writer.store_value(objective_np)
-            objective_nd.backward()
+                self._debug_writer.store_value(objective)
             if self.optimization_config.verbose:
-                msg_lst = ["[criterion = {}]".format(objective_np)]
+                msg_lst = ["[criterion = {}]".format(objective)]
                 for param, encoding in self.likelihood.param_encoding_pairs():
                     msg_lst.append(param_to_pretty_string(param, encoding))
                 logger.info('\n'.join(msg_lst))
-            return objective_np
-
-        arg_dict = {}
-        grad_dict = {}
-        params = self.likelihood.collect_params().values()
-        for param in params:
-            arg_dict[param.name] = param.data(ctx=self._ctx)
-            grad_dict[param.name] = param.grad(ctx=self._ctx)
+            return objective
 
-        return executor, arg_dict, grad_dict
+        return make_scipy_objective(executor), param_dict
 
     def fit(self, X, Y, profiler: GPMXNetSimpleProfiler = None):
```

Comment:
Do you think we may eventually be able to include this GPRegressor in autogluon-tabular as one of the supervised learning models for regression problems? Or are there are better 3rd party GP libraries out there? It would be nice to have at least 1 model for regression tasks that can easily produce both aleatoric+epistemic uncertainty estimates

Reply:
Nice idea.

Our code is pretty robust for what it is doing, but it is of course pretty limited, in the sense that we only implement a single kernel, and also only implement Gaussian noise.

Adding more kernels is not much of an issue, may do this as they are needed. Going to a likelihood different from Gaussian (important for binary or multi-class) is however quite tricky, and we have no plans to code this up (of course, if there was some volunteer, I'd be happy to advise).

One thing to be aware of is that GPs scale as O(n^3), while most of your methods probably scale as O(n).

I know Alex S is quite keen on GPyTorch, and this may be the best bet for a single GP package to be introduced to AG. It seems of high quality, and A Wilson is here to help I suppose.

The advantage of our code here is that it is so light, and it gets a few things right, so I'd for sure keep this in the core, and then I'd advice for GPyTorch (which has at least PyTorch dependence) for scalable GP models.

Reply:
With Alex S I mean Smola, not Shirkov